### ISSUE TITLE
Adding timeouts to tuntap_read/tuntap_write

### ISSUE THREAD

**Issue by cozis** [LaKabane/libtuntap#42]
Hi, I've been using libtuntap succesfully for one of my projects and am quite happy with it! Although I was thinking it would be nice to have the ability to express a timeout condition for reads and writes on the tun/tap device.

My idea was something like:
```
int tuntap_read(struct device *dev, void *buf, size_t size, int timeout_ms);
int tuntap_write(struct device *dev, void *buf, size_t size, int timeout_ms);
```
Users can default to the current behaviour by passing -1 to the newly added argument, or specify the maximum number of milliseconds that the call should block for.

This should be easy to implement on both windows and unix platforms. If this kind of change is welcome, I can start working on a proof of concept and we can go from there. 

Thanks for your attention.

**Comment by tleguern** [comment]
I am not against such a change, also I would probably prefer a separate function to set the timeout if it is not too cumbersome. What about the interaction with `tuntap_set_nonblocking()` too?

**Comment by tleguern** [comment]
Using `select()` is fine on Unix as long as the current interfaces are not simple wrappers on the new ones: if someone wants to implement more complex behaviors and use `epoll` or whatever they should be able to keep using the non-select interfaces.

Also can you change the `2` prefix to `_tm` or something capable of conveying the functionality?

**Comment by cozis** [comment]
Good idea. I'll restore the previous implementations and add a suffix for the new ones. 

On another note, should I use this thread or the one for the pull request?

**Comment by tleguern** [comment]
> On another note, should I use this thread or the one for the pull request?

This thread as this is already here that much of the conversation happened.

**Comment by cozis** [comment]
Hello! I restored the previous implementations and changed the suffix of the new functions.

Unfortunately the original version of `tuntap_write` and `tuntap_read` didn't work for me on windows because the `ReadFile` and `WriteFile` calls were done in non-overlapped mode while the descriptor was opened with `FILE_FLAG_OVERLAPPED`. Looks like this is a common problem and the commonly proposed solution is to just remove the flag. Since the new functions `tuntap_write_tm` and `tuntap_read_tm` require the overlapped mode, instead of removing the flag I modified the old implementations to also start in overlapped mode but block using `GetOverlappedResult`.

It may be worth noting that this would make it easy to implement `tuntap_set_nonblocking` on windows, which at the moment is not implemented. This would be done by only calling `GetOverlappedResult` in the read and write routines if a given "is_blocking" flag is true.

Let me know if this works for you!

**Comment by cozis** [comment]
> It may be worth noting that this would make it easy to implement tuntap_set_nonblocking on windows, which at the moment is not implemented. This would be done by only calling GetOverlappedResult in the read and write routines if a given "is_blocking" flag is true.

After thinking about it some more I realized that it probably wouldn't be so easy since in windows you can't just poll for the readiness of a descriptor like you can do in unix. In windows you can poll for data to be read, but only after having starting a read operation
